WIP: Use a tagged hash in bip-schnorr nonce derivation

jonasnick opened the pull request:
The currently specified nonce derivation algorithm is sufficiently obvious that someone may use it outside of a bip-schnorr implementation, for example in ECDSA. This could have catastrophic consequences and therefore it would be better to add a context dependent tag name to avoid collisions.

WIP because test vectors still need to be adjusted.

real-or-random commented on bip-schnorr.mediawiki:
```diff
@@ -110,6 +110,11 @@ The following convention is used, with constants as defined for secp256k1:
 *** Return the unique point ''P'' such that ''x(P) = x'' and ''y(P) = y''.
 ** The function ''point(x)'', where ''x'' is a 32-byte array, returns the point ''P = lift_x(int(x))''.
 ** The function ''hash(x)'', where ''x'' is a byte array, returns the 32-byte SHA256 hash of ''x''.
```
Now with both `hash` and `hash_tag`, I thin renaming `hash` to `sha256` is a good idea. It's just clearer.
(Maybe I said this already but I can't remember. Feel free to ignore if there were good arguments against it and we discussed this already...)

jonasnick replied:
removed `hash`

jonasnick commented:
I added a commit that improves the rationale (moved to design section, mentions that there's no guarantee for conflict-freeness in general, added examples) and also uses tagged hashing in the signature hash.

jonasnick commented:
Also removed WIP status

jonasnick commented on bip-taproot.mediawiki:
```diff
@@ -50,15 +50,8 @@ Not included in this proposal are additional features like new sighash modes or
-* Because the prefix ''SHA256(tag) || SHA256(tag)'' is a 64-byte long context-specific constant, optimized implementations are possible (identical to SHA256 itself, but with a modified initial state).
-* Using SHA256 of the tag name itself is reasonably simple and efficient for implementations that don't choose to use the optimization above.
+This includes the notion of tagged hashes which make collisions with existing uses of hash functions in Bitcoin infeasible.
+The reason is that so far nowhere in the Bitcoin protocol are hashes used where the input of SHA256 starts with two (non-double) SHA256 hashes.
```
The reason for collisions with existing uses of hash functions in Bitcoin being infeasible.

jonasnick commented:
@sipa better?
